### DISCUSSION TITLE
prisma on docker

### DISCUSSION THREAD

**Discussion by aisartag** [prisma/prisma#23134] [Q&A]
### Bug description

nestjs prisma create container docker  @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: 
- Database: 
- Node.js version: 

### Prisma Version

```

```

**Comment by nurul3101** [comment]
Hello @aisartag 👋 

Could you please share your Dockerfile?
How are you building your app?
Is `npx prisma generate` command a part of your docker file setup while building?

**Reply by aisartag** [reply]
# syntax=docker/dockerfile:1

# Comments are provided throughout this file to help you get started.
# If you need more help, visit the Dockerfile reference guide at
# https://docs.docker.com/go/dockerfile-reference/

# Want to help us make this template better? Share your feedback here: https://forms.gle/ybq9Krt8jtBL3iCk7

ARG NODE_VERSION=20.11.0

################################################################################
# Use node image for base image for all stages.
FROM node:${NODE_VERSION}-alpine as base

# Set working directory for all build stages.
WORKDIR /usr/src/app


################################################################################
# Create a stage for installing production dependecies.
FROM base as deps

# Download dependencies as a separate step to take advantage of Docker's caching.
# Leverage a cache mount to /root/.yarn to speed up subsequent builds.
# Leverage bind mounts to package.json and yarn.lock to avoid having to copy them
# into this layer.
RUN --mount=type=bind,source=package.json,target=package.json \
    --mount=type=bind,source=yarn.lock,target=yarn.lock \
    --mount=type=cache,target=/root/.yarn \
    yarn install --production --frozen-lockfile

################################################################################
# Create a stage for building the application.
FROM deps as build

# Download additional development dependencies before building, as some projects require
# "devDependencies" to be installed to build. If you don't need this, remove this step.
RUN --mount=type=bind,source=package.json,target=package.json \
    --mount=type=bind,source=yarn.lock,target=yarn.lock \
    --mount=type=cache,target=/root/.yarn \
    yarn install --frozen-lockfile 


# Copy the rest of the source files into the image.
COPY . .
# Run the build script.


COPY prisma ./prisma/

RUN yarn prisma db push
RUN yarn prisma generate


RUN yarn run build

################################################################################
# Create a new stage to run the application with minimal runtime dependencies
# where the necessary files are copied from the build stage.
FROM base as final

# Use production node environment by default.
ENV NODE_ENV production

# Run the application as a non-root user.
USER node

# Copy package.json so that package manager commands can be used.
COPY package.json .


# Copy the production dependencies from the deps stage and also
# the built application from the build stage into the image.
COPY --from=deps /usr/src/app/node_modules ./node_modules
COPY --from=build /usr/src/app/dist ./dist


# Expose the port that the application listens on.
EXPOSE 3001


# Run the application.
CMD yarn start:prod

**Reply by ludralph** [reply]
Hi @aisartag 

You're already running `yarn prisma generate` in your build stage, which is correct. However, the generated Prisma client might not be available in your final stage.

- Copying Prisma Files: You need to ensure that the generated Prisma client is copied to your final stage. Add the following line to your Dockerfile in the final stage:
```
COPY --from=build /usr/src/app/node_modules/.prisma ./node_modules/.prisma
```

- Environment Variables: Make sure you're providing the necessary environment variables, especially the `DATABASE_URL`. Add this to your final stage:
```
ENV DATABASE_URL="your_database_url_here"
```

- Prisma Schema: Copy your Prisma schema to the final stage:
```
COPY --from=build /usr/src/app/prisma ./prisma
```
- Running Prisma Generate in Final Stage: To ensure the Prisma client is generated for the correct environment, you might want to run prisma generate in the final stage:
```
RUN yarn prisma generate
```

Here's a modified version of your Dockerfile incorporating these changes:
```docker
# ... (previous stages remain the same)

FROM base as final

# Use production node environment by default.
ENV NODE_ENV production

# Copy package.json so that package manager commands can be used.
COPY package.json .

# Copy the production dependencies from the deps stage and also
# the built application from the build stage into the image.
COPY --from=deps /usr/src/app/node_modules ./node_modules
COPY --from=build /usr/src/app/dist ./dist
COPY --from=build /usr/src/app/node_modules/.prisma ./node_modules/.prisma
COPY --from=build /usr/src/app/prisma ./prisma

# Set your database URL
ENV DATABASE_URL="your_database_url_here"

# Run Prisma generate
RUN yarn prisma generate

# Expose the port that the application listens on.
EXPOSE 3001

# Run the application as a non-root user.
USER node

# Run the application.
CMD yarn start:prod
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
